File: Solution Union Find.cpp

#include <array>
#include <string>
#include <vector>
using namespace std;

/*
  Abridged Version of Union Find.
  Join by rank is not necessary for this particular problem.
 */
class Solution {
    
    inline static const int ALPHABET_SIZE = 26;
    array<int, ALPHABET_SIZE> rootsOfEqualElements;
    
public:
    bool equationsPossible(vector<string>& equations) {
        iota(rootsOfEqualElements.begin(), rootsOfEqualElements.end(), 0);

        for (const auto& equation : equations) {
            if (equation[1] == '=') {
                int rootFirst = findRoot(equation[0] - 'a');
                int rootSecond = findRoot(equation[3] - 'a');
                rootsOfEqualElements[rootFirst] = rootsOfEqualElements[rootSecond];
            }
        }

        for (const auto& equation : equations) {
            if (equation[1] == '!') {
                int rootFirst = findRoot(equation[0] - 'a');
                int rootSecond = findRoot(equation[3] - 'a');
                if (rootFirst == rootSecond) {
                    return false;
                }
            }
        }

        return true;
    }

private:
    int findRoot(int index) {
        if (rootsOfEqualElements[index] != index) {
            rootsOfEqualElements[index] = findRoot(rootsOfEqualElements[index]);
        }
        return rootsOfEqualElements[index];
    }
};

File: Solution Union Find.js

/*
 Abridged Version of Union Find.
 Join by rank is not necessary for this particular problem.
 */
/**
 * @param {string[]} equations
 * @return {boolean}
 */
var equationsPossible = function (equations) {
    const ALPHABET_SIZE = 26;
    const asciiCodeFor_a = 97;
    this.rootsOfEqualElements = Array.from(new Array(ALPHABET_SIZE).keys());

    for (let equation of equations) {
        if (equation.charAt(1) === '=') {
            let rootFirst = findRoot(equation.codePointAt(0) - asciiCodeFor_a);
            let rootSecond = findRoot(equation.codePointAt(3) - asciiCodeFor_a);
            rootsOfEqualElements[rootFirst] = rootsOfEqualElements[rootSecond];
        }
    }

    for (let equation of equations) {
        if (equation.charAt(1) === '!') {
            let rootFirst = findRoot(equation.codePointAt(0) - asciiCodeFor_a);
            let rootSecond = findRoot(equation.codePointAt(3) - asciiCodeFor_a);
            if (rootFirst === rootSecond) {
                return false;
            }
        }
    }

    return true;
};


/**
 * @param {number} index
 * @return {number}
 */
function findRoot(index) {
    if (this.rootsOfEqualElements[index] !== index) {
        this.rootsOfEqualElements[index] = findRoot(this.rootsOfEqualElements[index]);
    }
    return this.rootsOfEqualElements[index];
}
